File: navigator/path.py
from __future__ import annotations

from collections import deque
from collections.abc import Iterable, Mapping

from .models import Course


class CycleError(ValueError):
    """Raised when the course catalogue contains circular dependencies."""


def collect_required_courses(
    target_ids: Iterable[str],
    course_map: Mapping[str, Course],
    completed_ids: Iterable[str] | None = None,
) -> set[str]:
    """Return the ids required to reach all targets starting from completed ids."""

    completed = set(completed_ids or [])
    required: set[str] = set()

    def dfs(course_id: str) -> None:
        if course_id in required or course_id in completed:
            return
        required.add(course_id)
        for prereq in course_map[course_id].prerequisites:
            dfs(prereq)

    for target_id in target_ids:
        if target_id not in course_map:
            raise KeyError(f"Unknown course id '{target_id}'")
        dfs(target_id)

    return required


def build_learning_path(
    target_ids: Iterable[str],
    course_map: Mapping[str, Course],
    completed_ids: Iterable[str] | None = None,
) -> list[Course]:
    """Return an ordered list of courses needed to reach the selected targets."""

    completed = set(completed_ids or [])
    required_ids = collect_required_courses(target_ids, course_map, completed)
    ordered_ids = _topological_sort(required_ids, course_map, completed)
    return [course_map[course_id] for course_id in ordered_ids]


def _topological_sort(
    required_ids: set[str],
    course_map: Mapping[str, Course],
    completed_ids: set[str],
) -> list[str]:
    """Return a topologically sorted list of the required course ids."""

    in_degree: dict[str, int] = {}
    dependents: dict[str, set[str]] = {}

    for course_id in required_ids:
        course = course_map[course_id]
        filtered_prereqs = [
            prereq
            for prereq in course.prerequisites
            if prereq in required_ids and prereq not in completed_ids
        ]
        in_degree[course_id] = len(filtered_prereqs)
        for prereq in filtered_prereqs:
            dependents.setdefault(prereq, set()).add(course_id)

    queue = deque(course_id for course_id, degree in in_degree.items() if degree == 0)
    ordered: list[str] = []
    processed: set[str] = set()

    while queue:
        course_id = queue.popleft()
        if course_id in processed:
            continue
        processed.add(course_id)
        if course_id not in completed_ids:
            ordered.append(course_id)
        for dependent_id in dependents.get(course_id, set()):
            in_degree[dependent_id] -= 1
            if in_degree[dependent_id] == 0:
                queue.append(dependent_id)

    if len(processed) != len(required_ids):
        raise CycleError("Detected a cycle in the course prerequisites")

    return ordered

File: web/src/path.ts
import type { Course } from './types'

export class CycleError extends Error {}

const CASCADE_COMPLETION_IDS = new Set(['fachausbildung_wasserrettungsdienst'])

function expandCompleted(
  completedIds: Iterable<string>,
  courseMap: Map<string, Course>,
): Set<string> {
  const expanded = new Set<string>()
  const stack = [...completedIds]
  while (stack.length > 0) {
    const id = stack.pop()!
    if (expanded.has(id)) continue
    expanded.add(id)
    const course = courseMap.get(id)
    if (!course) continue
    if (CASCADE_COMPLETION_IDS.has(id)) stack.push(...course.prerequisites)
  }
  return expanded
}

export function collectRequiredCourses(
  targetIds: Iterable<string>,
  courseMap: Map<string, Course>,
  completedIds: Iterable<string> = [],
): Set<string> {
  const completed = expandCompleted(completedIds, courseMap)
  const required = new Set<string>()

  function dfs(id: string) {
    if (required.has(id) || completed.has(id)) return
    required.add(id)
    const course = courseMap.get(id)
    if (!course) throw new Error(`Unbekannter Lehrgang '${id}'`)
    for (const prereq of course.prerequisites) dfs(prereq)
  }

  for (const id of targetIds) dfs(id)
  return required
}

export function buildLearningPath(
  targetIds: Iterable<string>,
  courseMap: Map<string, Course>,
  completedIds: Iterable<string> = [],
): Course[] {
  const completed = expandCompleted(completedIds, courseMap)
  const requiredIds = collectRequiredCourses(targetIds, courseMap, completed)
  const orderedIds = topologicalSort(requiredIds, courseMap, completed)
  return orderedIds.map(id => courseMap.get(id)!)
}

function topologicalSort(
  requiredIds: Set<string>,
  courseMap: Map<string, Course>,
  completedIds: Set<string>,
): string[] {
  const inDegree = new Map<string, number>()
  const dependents = new Map<string, Set<string>>()

  for (const id of requiredIds) {
    const course = courseMap.get(id)!
    const filteredPrereqs = course.prerequisites.filter(
      p => requiredIds.has(p) && !completedIds.has(p),
    )
    inDegree.set(id, filteredPrereqs.length)
    for (const p of filteredPrereqs) {
      if (!dependents.has(p)) dependents.set(p, new Set())
      dependents.get(p)!.add(id)
    }
  }

  const queue: string[] = []
  for (const [id, deg] of inDegree) {
    if (deg === 0) queue.push(id)
  }

  const ordered: string[] = []
  const processed = new Set<string>()

  while (queue.length > 0) {
    const id = queue.shift()!
    if (processed.has(id)) continue
    processed.add(id)
    if (!completedIds.has(id)) ordered.push(id)
    for (const depId of dependents.get(id) ?? []) {
      const deg = inDegree.get(depId)! - 1
      inDegree.set(depId, deg)
      if (deg === 0) queue.push(depId)
    }
  }

  if (processed.size !== requiredIds.size) {
    throw new CycleError('Zyklus in den Voraussetzungen erkannt')
  }
  return ordered
}
